esp-hal/src/sdm.rs

```diff
@@ -0,0 +1,296 @@
+
+use core::marker::PhantomData;
+
+use fugit::HertzU32;
+
+use crate::{
+    clock::Clocks,
+    gpio::{OutputPin, OutputSignal},
+    peripheral::{Peripheral, PeripheralRef},
+    peripherals::GPIO_SD,
+    private,
+};
+
+/// Sigma-Delta modulation peripheral driver.
+pub struct Sdm<'d> {
+    /// Channel 0
+    pub channel0: ChannelRef<'d, 0>,
+
+    /// Channel 1
+    pub channel1: ChannelRef<'d, 1>,
+
+    /// Channel 2
+    pub channel2: ChannelRef<'d, 2>,
+
+    /// Channel 3
+    pub channel3: ChannelRef<'d, 3>,
+
+    #[cfg(any(esp32, esp32s2, esp32s3))]
+    /// Channel 4
+    pub channel4: ChannelRef<'d, 4>,
+
+    #[cfg(any(esp32, esp32s2, esp32s3))]
+    /// Channel 5
+    pub channel5: ChannelRef<'d, 5>,
+
+    #[cfg(any(esp32, esp32s2, esp32s3))]
+    /// Channel 6
+    pub channel6: ChannelRef<'d, 6>,
+
+    #[cfg(any(esp32, esp32s2, esp32s3))]
+    /// Channel 7
+    pub channel7: ChannelRef<'d, 7>,
+}
+
+/// Channel errors
+#[derive(Debug, Clone, Copy, PartialEq)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub enum Error {
+    /// Prescale out of range
+    PrescaleRange,
+}
+
+impl<'d> Drop for Sdm<'d> {
+    fn drop(&mut self) {
+        GPIO_SD::enable_clock(false);
+    }
+}
+
+impl<'d> Sdm<'d> {
+    /// Initialize driver using a given SD instance.
+    pub fn new(_sd: impl crate::peripheral::Peripheral<P = GPIO_SD> + 'd) -> Self {
+        GPIO_SD::enable_clock(true);
+
+        Self {
+            channel0: ChannelRef::new(),
+            channel1: ChannelRef::new(),
+            channel2: ChannelRef::new(),
+            channel3: ChannelRef::new(),
+
+            #[cfg(any(esp32, esp32s2, esp32s3))]
+            channel4: ChannelRef::new(),
+            #[cfg(any(esp32, esp32s2, esp32s3))]
+            channel5: ChannelRef::new(),
+            #[cfg(any(esp32, esp32s2, esp32s3))]
+            channel6: ChannelRef::new(),
+            #[cfg(any(esp32, esp32s2, esp32s3))]
+            channel7: ChannelRef::new(),
+        }
+    }
+}
+
+/// Sigma-Delta modulation channel reference.
+pub struct ChannelRef<'d, const N: u8> {
+    _phantom: PhantomData<&'d ()>,
+}
+
+impl<'d, const N: u8> ChannelRef<'d, N> {
+    fn new() -> Self {
+        Self {
+            _phantom: PhantomData,
+        }
+    }
+
+    /// Configure and connect sigma-delta channel to output
+    pub fn connect<O: OutputPin>(
+        &'d self,
+        output: impl Peripheral<P = O> + 'd,
+        frequency: HertzU32,
+    ) -> Result<Channel<'d, N, O>, Error> {
+        crate::into_ref!(output);
+
+        let signal = CHANNELS[N as usize];
+
+        output.connect_peripheral_to_output(signal, private::Internal);
+
+        let channel = Channel { _ref: self, output };
+
+        channel.set_frequency(frequency)?;
+
+        Ok(channel)
+    }
+}
+
+/// Sigma-Delta modulation channel handle.
+pub struct Channel<'d, const N: u8, O: OutputPin> {
+    _ref: &'d ChannelRef<'d, N>,
```

Comment:
Using `'d` for both the borrow and the `ChannelRef` will have some awkward consequences around static.

The `Channel` should just take ownership of the `ChannelRef`, at least then the `Channel`s can be sent off to other threads.

Reply:
Of course, I trying to redesign that.